app/assets/javascripts/models/beta/iteration.js

```diff
@@ -156,52 +196,54 @@ const pastAverageVelocity = (pastIterations) => {
 }
 
 const addToSprintFromBacklog = (sprints, project, story, velocity) => {
-  const sprintIndex = sprints.length && sprints.length - 1;
-  const hasSpace = canTakeStory(sprints[sprintIndex], Story.getPoints(story));
-
-  if (hasSpace) {
-    return addStoryToSprint(project, sprints, sprintIndex, story, velocity);
-  }
-
   const iterationNumber = getNextIterationNumber(sprints);
 
-  sprints[sprintIndex + 1] = createSprint(
+  return [...sprints, createSprint(
     iterationNumber,
     getDateForIterationNumber(iterationNumber, project),
-    undefined,
-    velocity
-  );
-
-  return addStoryToSprint(project, sprints, sprintIndex + 1, story, velocity);
+    false,
+    velocity,
+    [story],
+    Story.getPoints(story)
+  )];
 };
 
 export const groupBySprints = (stories = [], project, initialSprintNumber = 1, pastIterations) => {
   const velocity = sprintVelocity(project, pastIterations)
 
   return stories.reduce((sprints, story) => {
     const firstSprintIndex = 0;
-    const isFromSprintInProgress = !Story.isUnstarted(story);
-
-    sprints = handleSprintsOverflow(
-      project,
-      sprints,
-      Story.getPoints(story),
-      initialSprintNumber,
-      velocity
-    );
+    const isFromSprintInProgress = Story.isInProgress(story);
+
+    if (causesOverflow(story, velocity) && !isFromSprintInProgress) {
+      return addSprintWithOverflow(project, sprints, story, velocity, initialSprintNumber);
+    }
 
     if (sprints.length === 0) {
-      createFirstSprint(sprints, project, initialSprintNumber, velocity);
+      sprints = [createFirstSprint(project, initialSprintNumber, velocity)];
     }
 
     if (isFromSprintInProgress) {
       return addStoryToSprint(project, sprints, firstSprintIndex, story, velocity);
     }
 
+    if (canEnterLastAddedSprint(sprints, story, velocity)) {
+      return addStoryToSprint(project, sprints, sprints.length - 1, story, velocity);
+    }
+
     return addToSprintFromBacklog(sprints, project, story, velocity);
   }, []);
 };
 
+const canEnterLastAddedSprint = (sprints, story, velocity) => {
```

Comment:
This function is using `velocity` param?
Also, you could rewrite the `return` logic since you need `lastSprint` and the return of `canTakeStory` to be `true`:
```diff
- return lastSprint ? canTakeStory(lastSprint, storyPoints) : false;
+ return lastSprint && canTakeStory(lastSprint, storyPoints);
```

Reply:
Yes, you're right, the `velocity` was not being used. Thanks